gh-121040: Use __attribute__((fallthrough))

vstinner opened the pull request:
Annotate explicitly "fall through" switch cases with a new _Py_FALLTHROUGH macro which uses __attribute__((fallthrough)) if available.

Fix warnings when using -Wimplicit-fallthrough compiler flag.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-121040 -->
* Issue: gh-121040
<!-- /gh-issue-number -->

vstinner committed:
gh-121040: Use __attribute__((fallthrough))

Annotate explicitly "fall through" switch cases with a new
_Py_FALLTHROUGH macro which uses __attribute__((fallthrough)) if
available.

Fix warnings when using -Wimplicit-fallthrough compiler flag.

sobolevn commented on Include/pyport.h:
```diff
@@ -607,4 +607,19 @@ extern "C" {
 #endif
 
+// Explicit fallthrough in switch case.
+// Example:
```
Maybe also mention `-Wimplicit-fallthrough`?

sobolevn reviewed (commented):
Looks like you need to run `make regen-cases` (or maybe even `make regen-all`) :)

vstinner committed:
Run make regen-cases

vstinner committed:
mention -Wimplicit-fallthrough in _Py_FALLTHROUGH doc

AlexWaygood commented on Include/pyport.h:
```diff
@@ -607,4 +607,21 @@ extern "C" {
+//     }
+//
+// __attribute__((fallthrough)) was introduced in GCC 7.
+#if defined(__has_attribute) && __has_attribute(fallthrough)
```
it looks like there are many compiler warnings here on the GitHub Windows runners:

```
unexpected tokens following preprocessor directive - expected a newline [D:\a\cpython\cpython\PCbuild\_freeze_module.vcxproj]
```

corona10 replied:
Yeah I worried about this issue.

vstinner replied:
I rewrote the fix to avoid preprocessor issue with __has_attribute(). Instead, I added _Py__has_attribute() macro.

vstinner committed:
Add _Py__has_attribute() macro

No longer define __has_attribute() if it's not defined.

Move also _Py__has_builtin() at the top of pyport.h.

ericsnowcurrently commented on Python/pyhash.c:
```diff
@@ -442,13 +442,13 @@ siphash24(uint64_t k0, uint64_t k1, const void *src, Py_ssize_t src_sz) {
-        case 2: pt[1] = in[1]; /* fall through */
-        case 1: pt[0] = in[0]; /* fall through */
+        case 3: pt[2] = in[2]; _Py_FALLTHROUGH;
+        case 2: pt[1] = in[1]; _Py_FALLTHROUGH;
+        case 1: pt[0] = in[0]; break;
     }
```
Could this be sorted out using `default`?

```suggestion
        case 2: pt[1] = in[1]; _Py_FALLTHROUGH;
        case 1: pt[0] = in[0]; _Py_FALLTHROUGH;
        default: /* Do nothing. */
    }
```

That would also make it more explicit.  (Also I thought there is a compiler warning for no having `default`.)  Then again, your change using `break` is probably fine given the numerical progression of the case values.

vstinner replied:
Adding `default:` case would change the generated machine code and can make it less efficient. There is nothing to do for case 0 for example. This code is performance sensitive. `break;` is just a standard for switch/case.

vstinner replied:
Also, I'm the one who added the `/* fall through */` comment by mistake in 2017 :-) https://github.com/python/cpython/commit/c0e77364ca29df6cfb311e79892955c92bd8e595

ericsnowcurrently reviewed (approved):
LGTM

I left one small comment where you might consider a slightly different spelling, but I'll leave that up to you.

vstinner committed:
Fix typo in the doc

vstinner committed:
Merge branch 'main' into fallthrough
